### DISCUSSION TITLE
Edge Function size limit exceeded when used with NextAuth and Vercel

### DISCUSSION THREAD

**Discussion by taylor-lindores-reeves** [resend/react-email#1707] [General]
I'm encountering an issue when using react-email in conjunction with NextAuth on a Vercel deployment. The Edge Function for my authentication route is exceeding Vercel's size limit. Here are the details:

```
import { env } from "@/env.mjs";
import { normalizeIdentifier } from "@/utils/normalize-identifier";
import type { User as UserType } from "@prisma/client";
import type { NextAuthConfig } from "next-auth";
import { decode, encode } from "next-auth/jwt";
import type { Provider } from "next-auth/providers";
import ResendProvider from "next-auth/providers/resend";
import "server-only";

const providers: Provider[] = [
	ResendProvider({
		apiKey: env.RESEND_API_KEY,
		server: env.RESEND_SERVER,
		from: env.RESEND_FROM_ADDRESS,
		normalizeIdentifier: normalizeIdentifier,
		sendVerificationRequest: async (params) => {
			const resend = await import("@/lib/resend").then((mod) => mod.resend);

			const { identifier: to, provider, url } = params;
			// if the to email address does not match the domain, throw an error
			const VerifyRequestEmailTemplate = await import(
				"@/emails/templates/verify-request"
			).then((mod) => mod.default);

			const { error } = await resend.emails.send({
				from: provider.from ?? "",
				to,
				subject: `Sign in to ${env.NEXT_PUBLIC_APP_DOMAIN}`,
				react: VerifyRequestEmailTemplate({ data: { url } }),
			});

			if (error) throw new Error(`Resend ${error.name}: ${error.message}`);
		},
	}),
];

export default {
	providers,
	secret: env.NEXTAUTH_SECRET,
	trustHost: true,
	debug: false,
	session: { strategy: "jwt" },
	jwt: { encode, decode },
	pages: { signIn: "/auth/signin" },
	callbacks: {
		// Logged in users are authenticated, otherwise redirect to login page
		authorized: async ({ auth }) => !!auth?.user,
		async signIn({ user }) {
			if (user) return true;
			return false;
		},
		async session({ session, token }) {
			if (session.user) {
				session.user = token?.user as UserType;
			}
			return session;
		},
		async jwt({ token, user, trigger }) {
			if (user) {
				if (trigger === "signIn") {
					token.user = user;
					if (user?.emailVerified && !user?.stripeCustomerId) {
						const {
							getStripeCustomerId,
							updateStripeCustomer,
							assignHistoricalBookings,
						} = await import("@/utils/stripe-server-utils");

						const stripeCustomerId = await getStripeCustomerId(
							user.email as string,
						);

						token.user = {
							...user,
							stripeCustomerId,
						};

						await updateStripeCustomer(user as UserType, stripeCustomerId);
						await assignHistoricalBookings(user as UserType, stripeCustomerId);
					}
				}
			}

			return token;
		},
	},
} satisfies NextAuthConfig;

```

- Next.js version: 15.0.0-canary.167
- react-email version: 0.0.25
- NextAuth version: 5.0.0-beta.21
- Deployment platform: Vercel

**Description:**
When deploying my Next.js application to Vercel, the Edge Function for the NextAuth route is exceeding the 2 MB size limit. After investigation, it appears that react-email and its dependencies are contributing significantly to this bundle size.

`Error: The Edge Function "api/auth/[...nextauth]" size is 2.17 MB and your plan size limit is 2 MB.`

**Steps to Reproduce:**
1. Set up a Next.js project with NextAuth
2. Implement react-email for email templates
3. Configure the auth API route to use the Edge runtime
4. Deploy to Vercel

**Expected Behavior:**
The Edge Function should be minimal size within the 2 MB size limit to allow successful deployment on Vercel.

**Actual Behavior:**
The Edge Function size is 2.17 MB, exceeding the limit and preventing deployment.

**Questions:**
1. Are there any known issues or recommendations for using react-email in Edge Functions?
2. Are there any optimisations or configurations we can apply to reduce the bundle size when using react-email with NextAuth in an Edge environment?
3. Do you have any plans to optimise react-email for Edge Function compatibility?

Any insights or suggestions would be greatly appreciated.

**Comment by taylor-lindores-reeves** [comment]
<img width="2544" alt="Screenshot 2024-09-30 at 08 00 47" src="https://github.com/user-attachments/assets/c02efcbf-4c42-438e-9522-3a86a5a78313">

Also, there is no way to run the build CLI command. It either deletes the .react-email directory I created, or when running the build command without a .react-email directory:

`Starting build process...Error: ENOENT: no such file or directory, open '/path/to/my/project/.react-email/next.config.js'`

**Comment by taylor-lindores-reeves** [comment]
@zenorocha wondering if this is a known issue, or a limitation with the combination of using React Email & Vercel infra & Edge Config AuthJS?

**Comment by gabrielmfern** [comment]
Is your template using Tailwind? Would be interesting for you to test this out on the latest canaries, since we have just released #1383

**Reply by taylor-lindores-reeves** [reply]
I've tested it out. Unfortunately I still get a 2mb limit (the bundle size for the [...nextauth] function hasn't changed).

This led me to believe the issue is actually with AuthJS. Removing the `export const runtime = 'edge'` config from [...nextauth]/route.ts worked and no issues since.

The bundle size is definitely a bit smaller now in the canary version, it just didn't fix the issue.
